### DISCUSSION TITLE
How to best handle/navigate optional arguments in a signature?

### DISCUSSION THREAD

**Discussion by windowsrefund** [saghen/blink.cmp#815] [Q&A]
Greets. I find myself somewhat confused over what I believe is the signature and what my expectations around its behavior should be. I'll demonstrate with something simple and common. Assuming I'm working with Lua and type `vim.keymap.de` and then accept the `del` suggestion, I now see the following:

`vim.keymap.del(modes, lhs, opts?)`

Given the first 2 arguments are mandatory, I now move to the first, type my string, move to the next arg, type my string.... Now what? I'm wondering what the fastest and cheapest way might be to leave well enough alone at this point? Seems like the things I've tried have left me with the useless `opts?` as the final argument which I then have to go "clean up". Then there's the issue of removing the needless comma after the 2nd arg. Granted, a format would take care of that but in general, I feel like I'm hitting a wall as soon as I complete that 2nd arg (in this example).

So what do the Jedi do here? Feel like I'm missing something really obvious.

**Comment by memchr** [comment]
It is a feature in lua-ls called callSnippet, which essentially replaces a call kind with a snippet. There is an
[ongoing issue](https://github.com/LuaLS/lua-language-server/issues/1699) requesting the implementation of a feature that makes the generated snippet include only the mandatory arguments.

You can work around this by disabling callSnippet entirely, since Blink can [automatically add brackets](https://cmp.saghen.dev/configuration/completion.html#auto-brackets) to calls without LSP support, and you can enable [signature help](https://cmp.saghen.dev/configuration/signature.html) to compensate for the lost information

```lua
-- lspconfig
lspconfig.lua_ls.setup {
    settings = {
        Lua = {
            completion = {
                callSnippet = "Disable",
            }
        }
    }
}

-- blink
require("blink.cmp").setup{
    signature = {
        enabled = true,
    },
}
```
